nixos/modules/services/databases/redis.nix

```diff
@@ -295,6 +295,32 @@ in
         StateDirectoryMode = "0700";
         # Access write directories
         UMask = "0077";
+        # Capabilities
+        CapabilityBoundingSet = "";
+        # Security
+        NoNewPrivileges = true;
+        # Sandboxing
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateUsers = true;
+        ProtectClock = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectControlGroups = true;
+        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        PrivateMounts = true;
+        # System Call Filtering
+        SystemCallArchitectures = "native";
+        SystemCallFilter = "~@clock @cpu-emulation @debug @keyring @memlock @module @mount @obsolete @privileged @raw-io @reboot @resources @setuid @swap";
```

Comment:
I remember Redis sometimes complains that it is not happy about the default process limits defaults/file descriptor limtis. Does it try to change those on its own (i.e. `@resources` would no longer allow that) on startup? A test would be `strace -e setrlimit redis-server`
I think we could also increase the values to the recommended ones by upstream...

Reply:
I can't check.
```diff --git a/nixos/modules/services/databases/redis.nix b/nixos/modules/services/databases/redis.nix
index 24fe4ab3cc2..46d021209f8 100644
--- a/nixos/modules/services/databases/redis.nix
+++ b/nixos/modules/services/databases/redis.nix
@@ -282,7 +282,7 @@ in
       '';

       serviceConfig = {
-        ExecStart = "${cfg.package}/bin/redis-server /run/redis/redis.conf";
+        ExecStart = "${pkgs.strace}/bin/strace -e setrlimit -ff -o /var/lib/redis/trace ${cfg.package}/bin/redis-server /run/redis/redis.conf";
         Type = "notify";
         # User and group
         User = "redis";
```
Result:
```
redis.service: Failed with result 'core-dump'.
```

Reply:
https://github.com/redis/redis/blob/6.2.1/src/server.c#L2899-L2972 - It seems that this check?

Reply:
Without this PR error increase file descriptor limits.
```
You requested maxclients of 680000 requiring at least 680032 max file descriptors.
Server can't set maximum open files to 680032 because of OS error: Operation not permitted.
Current maximum open files is 524288. maxclients has been reduced to 524256 to compensate for low ulimit. If you need higher maxclients increase 'ulimit -n'.

```

Reply:
You can also run redis with strace from commandline without systemd. Systemd kills strace because `SystemCallFilter` disables `@debug`, which includes ptrace(2) as used by strace. Maybe `serviceConfig.LimitNOFILE = 680032;` in the systemd service could be set to shut up the message?

Reply:
Adding the parameter `serviceConfig.LimitNOFILE = 680032;` fixes the error

Reply:
> Systemd kills strace because `SystemCallFilter` disables `@debug`, which includes ptrace(2) as used by strace.

Checked without sandbox mode.

Reply:
Working with this patch:
```diff --git a/nixos/modules/services/databases/redis.nix b/nixos/modules/services/databases/redis.nix
index 24fe4ab3cc2..5c96caf50d4 100644
--- a/nixos/modules/services/databases/redis.nix
+++ b/nixos/modules/services/databases/redis.nix
@@ -296,7 +296,8 @@ in
         # Access write directories
         UMask = "0077";
         # Capabilities
-        CapabilityBoundingSet = "";
+        AmbientCapabilities = [ "CAP_SYS_RESOURCE" ];
+        CapabilityBoundingSet = [ "CAP_SYS_RESOURCE" ];
         # Security
         NoNewPrivileges = true;
         # Sandboxing
@@ -304,7 +305,7 @@ in
         ProtectHome = true;
         PrivateTmp = true;
         PrivateDevices = true;
-        PrivateUsers = true;
+        PrivateUsers = false;
         ProtectClock = true;
         ProtectHostname = true;
         ProtectKernelLogs = true;
```

Reply:
Optimal variant inrease manual with `serviceConfig.LimitNOFILE`. Example - https://github.com/redis/redis-debian/blob/master/debian/redis-server.default

Reply:
draft:
```diff --git a/nixos/modules/services/databases/redis.nix b/nixos/modules/services/databases/redis.nix
index 24fe4ab3cc2..b88bffaedbd 100644
--- a/nixos/modules/services/databases/redis.nix
+++ b/nixos/modules/services/databases/redis.nix
@@ -4,6 +4,7 @@ with lib;

 let
   cfg = config.services.redis;
+  ulimit = $(cfg.maxclient + 32)

   mkValueString = value:
     if value == true then "yes"
@@ -121,6 +122,12 @@ in
         description = "Set the number of databases.";
       };

+      maxclients = mkOption {
+        type = types.int;
+        default = 10000;
+        description = "Set the max number of connected clients at the same time.";
+      };
+
       save = mkOption {
         type = with types; listOf (listOf int);
         default = [ [900 1] [300 10] [60 10000] ];
@@ -253,6 +260,7 @@ in
         logfile = cfg.logfile;
         syslog-enabled = cfg.syslog;
         databases = cfg.databases;
+        maxclients = cfg.maxclients;
         save = map (d: "${toString (builtins.elemAt d 0)} ${toString (builtins.elemAt d 1)}") cfg.save;
         dbfilename = "dump.rdb";
         dir = "/var/lib/redis";
@@ -299,6 +307,8 @@ in
         CapabilityBoundingSet = "";
         # Security
         NoNewPrivileges = true;
+        # Process Properties
+        LimitNOFILE = ${ulimit};
         # Sandboxing
         ProtectSystem = "strict";
         ProtectHome = true;
```
I don't know how to increase the parameter by +32

Reply:
Fixed.